remove uses of errors.Is, which requires go1.13, move go1.16/go1.21 tests to separate file

thaJeztah opened the pull request:
relates to:

- https://github.com/spf13/pflag/pull/443
- https://github.com/spf13/pflag/pull/429


### remove uses of errors.Is, which requires go1.13

Commit 1bf832c6fec8a35a8c1d61e5fb14f5ce404197ef introduced the use of
`errors.Is`, which was added in [go1.13], but the `go.mod` was not
updated to reflect this, causing compile failures on go1.12:

    docker run -it --rm -v ./:/pflag -w /pflag golang:1.12 sh -c 'go test -v ./...'
    # github.com/spf13/pflag [github.com/spf13/pflag.test]
    ./flag.go:1190:7: undefined: errors.Is
    ./flag.go:1219:7: undefined: errors.Is
    ./bool_func_test.go:86:28: cannot use stdFSet (type *flag.FlagSet) as type BoolFuncFlagSet in argument to runCase:
        *flag.FlagSet does not implement BoolFuncFlagSet (missing BoolFunc method)
    ...

As the error that is tested will not be wrapped, we can omit the `errors.Is`,
and instead continue doing a straight comparison.

[go1.13]: https://pkg.go.dev/errors@go1.13#Is


### move Func, BoolFunc, tests as they require go1.21

Commit 69bc3bd5b37fa90e994be9acecf7430269591713 added support for Func()
and BoolFunc() to match stdlib. However, the Func method was added in
[go1.16.0], and BoolFunc in [go1.21.0], so running the tests on older
versions of Go would fail;

    docker run -it --rm -v ./:/pflag -w /pflag golang:1.21 sh -c 'go test -v ./...'
    # github.com/spf13/pflag [github.com/spf13/pflag.test]
    ./bool_func_test.go:86:28: cannot use stdFSet (type *flag.FlagSet) as type BoolFuncFlagSet in argument to runCase:
        *flag.FlagSet does not implement BoolFuncFlagSet (missing BoolFunc method)
    ./bool_func_test.go:113:21: undefined: io.Discard
    ./bool_func_test.go:116:28: cannot use stdFSet (type *flag.FlagSet) as type BoolFuncFlagSet in argument to runCase:
        *flag.FlagSet does not implement BoolFuncFlagSet (missing BoolFunc method)
    ./bool_func_test.go:139:7: undefined: errors.Is
    ./func_test.go:92:28: cannot use stdFSet (type *flag.FlagSet) as type FuncFlagSet in argument to runCase:
        *flag.FlagSet does not implement FuncFlagSet (missing Func method)
    ./func_test.go:119:21: undefined: io.Discard
    ./func_test.go:122:28: cannot use stdFSet (type *flag.FlagSet) as type FuncFlagSet in argument to runCase:
        *flag.FlagSet does not implement FuncFlagSet (missing Func method)
    ./func_test.go:145:7: undefined: errors.Is
    ./func_test.go:145:7: too many errors
    FAIL	github.com/spf13/pflag [build failed]

This patch moves the tests to a separate file that is not built for older
versions of Go.

[go1.16.0]: https://pkg.go.dev/flag@go1.16.0#Func
[go1.21.0]: https://pkg.go.dev/flag@go1.21.0#BoolFunc

tomasaschan commented:
Given the outrage at recent breaking changes being pushed without moving to v2, I don't think we can bump the required go version before that either. So whatever we have that requires later versions should either be rewritten to work on 1.12, or be moved into flags that can be feature toggled (like you did for the Func/BoolFunc stuff).

I have been meaning to start a conversation about what the path to 2.0 looks like, and what type of changes we want to get in there, but I don't think the right medium for such a discussion is currently available (which is one of the reasons I filed for widening my permissions).

thaJeztah commented:
OK 👍  I dropped the first commit, and replaced it with a commit that removes the `errors.Is` from "production" code. The other uses of `errors.Is` were in tests that are moved to a Go >= 1.21 file in the second commit.

tomasaschan reviewed (approved):
I wasn't able to make #449 target this PR (probably because it comes from a fork?) but it does verify that this works well on 1.12 too, so I'm gonna go with it. Holding off merging for a bit to see if we want to do a 1.0.10 with _only_ #447 before we merge anything else.

thaJeztah commented:
I think both https://github.com/spf13/pflag/pull/447 and this one would be good to include in v1.0.10, because this one fixes a bug where the `go.mod`'s requirements don't match the code.

That said; I wouldn't consider either to be _critical_, or at least, using go1.12 would be a very niche-case (it happens, but usually it would be in some internal codebases in companies maintaining their own LTS Go versions - I know Google itself did so for a long time.

I think it's fine to keep https://github.com/spf13/pflag/pull/449 open as draft; you can rebase it once this one's merged, but I can include it in my follow-up to address some linting issues.
